### ISSUE TITLE
[Discussion] Directed vs Undirected graphs in DGL

### ISSUE THREAD

**Issue by chaitjo** [dmlc/dgl#2198]
https://github.com/dmlc/dgl/blob/9488bace208911b1bc1efb17774d8000fed1acac/examples/pytorch/ogb/ogbn-arxiv/gcn.py#L241

Hi DGL team, I had a quick query re. directed and undirected graphs in DGL. In the example referenced above for OGBN-ARXIV, the graph is defined as a directed graph originally, but one needs to make it undirected by adding reverse edges in DGL. In fact, without adding these reverse edges, **I found performance to be very very poor regardless of architecture**.

1. Is using directed vs. undirected graphs a completely empirical choice in DGL when working with OGB, i.e. should I always try both and see the performance on a validation set to figure it out?
2. Do you recommend doing this reverse edges trick in other OGB datasets? For e.g., how about for graph classification using MOLPCBA; will it not impact the semantics of the task? Additionally, I can see this trick causing some data corruption issues when sampling training edges if I implement it in link prediction tasks such as COLLAB; what would you recommend?

Best,
Chaitanya

cc. @vijaydwivedi75, @weihua916 (not sure if this issue is better raised in OGB or in DGL).

**Comment by VoVAllen** [comment]
We also found convert graph to bidirected greatly improved the performance in ogb-paper100M. However, I believe this is due to limited research on the real directed graph. Most current work assumes graphs undirected. Maybe doing different message passing on the reversed graph could gain better performance. I believe this is a worth digging area

**Comment by chaitjo** [comment]
Interesting! I suppose this discussion is related to the idea that the graph given apriori may not be the best graph to use for message-passing, e.g. [this recent blog post](https://towardsdatascience.com/manifold-learning-2-99a25eeb677d).   

> Most current work assumes graphs undirected.

Indeed, for example, in case of molecular graphs. However, does DGL's decision to implement the undirected graphs via doubly linked graphs (i.e. with reverse edges) also not lead to issues in increased memory usage? 

Additionally, doesn't this style of adding reverse edges impact the semantics of the task in the case of e.g., OGBN-ARXIV? (Directed edges indicate papers citing each other.)

**Comment by jermainewang** [comment]
> I suppose this discussion is related to the idea that the graph given apriori may not be the best graph to use for message-passing,

They are indeed very related! There are a lot of noises in real world graph. Does an edge of a-user-clicking-an-item truly indicating the user's intention? Is the lack of a certain link a truly negative signal? Then does the message passing guided by those possibly misinterpreted edges a proper one? I don't have good answers to them. I think the current de-facto strategy of adding reverse edges is more based on human instincts.

> However, does DGL's decision to implement the undirected graphs via doubly linked graphs (i.e. with reverse edges) also not lead to issues in increased memory usage?

Unfortunately, it does lead to an increased memory usage. Currently DGL maintains both CSR and CSC formats for storing a graph to allow efficient operations on in-neighbors as well as out-neighbors. For undirected graphs, CSR == CSC so one format is enough. We haven't implemented this optimization and we are considering it in the next release.

> Additionally, doesn't this style of adding reverse edges impact the semantics of the task in the case of e.g., OGBN-ARXIV? (Directed edges indicate papers citing each other.)

I agree with you that this perhaps changes the semantics. But I'm curious on why the empirical results still get positive gain from doing so. To me, perhaps a more reasonable way is to treat the reverse edges as another type (e.g., cited-by).

**Comment by BarclayII** [comment]
My two cents as a supplement to @jermainewang : 

Adding reverse edges makes a lot of sense in terms of message passing, because we not only want to propagate messages from "citers" (i.e. ones that cite others) to "citees" (i.e. ones that are being cited), but also "citees" to "citers", since the fact that a paper is "being cited by another paper" can also be informative to the paper being considered.

It will probably be interesting to see whether treating "cites" and "cited-by" as two different types could bring more benefit.

**Comment by chaitjo** [comment]
Thank you for the insightful comments!

>It will probably be interesting to see whether treating "cites" and "cited-by" as two different types could bring more benefit.

Yes, this would be pretty intuitive.

**Comment by tkipf** [comment]
If you want to preserve the information coming from the directionality of an edge, a good solution is to assign a different edge type to inverse edges and to inform the GNN message passing updates about these edge types (e.g. a separate set of parameters for each edge type).

You can find an example of this inverse edge type approach in our R-GCN paper: https://arxiv.org/abs/1703.06103

**Comment by ketyi** [comment]
> If you want to preserve the information coming from the directionality of an edge, a good solution is to assign a different edge type to inverse edges and to inform the GNN message passing updates about these edge types (e.g. a separate set of parameters for each edge type).
> 
> You can find an example of this inverse edge type approach in our R-GCN paper: https://arxiv.org/abs/1703.06103

Hi Thomas @tkipf,

I have 2 questions:

1. You didn't have an ablation study about the need of reverse edges and I'm not aware of such a study. Are the models generally perform significantly worse without reverse edges?
2. I've found that DGL implementations seem to be not 100% conform with Figure 2 in your paper: https://discuss.dgl.ai/t/is-heterographconv-truly-compatible-with-the-r-gcn-paper/2800 Do you think I've mistaken something?

**Comment by BarclayII** [comment]
@ketyi I have replied in https://discuss.dgl.ai/t/is-heterographconv-truly-compatible-with-the-r-gcn-paper/2800.  Let's move our discussion about RGCN there.
